### ISSUE TITLE
Deprecating and removing `allow_multiple_sessions` feature?

### ISSUE THREAD

**Issue by larshesel** [vernemq/vernemq#1067]
We're considering the future of the `allow_multiple_sessions` feature and if we should deprecate it and eventually remove it in VerneMQ 2.0.

The `allow_multiple_sessions` feature is a non-standard feature which allows MQTT clients to have multiple connections/sessions using the same client-id and messages for which the clients have subscribed can be distributed between the clients sharing the client-id and in this way many clients can shared the load of the incoming messages.

However, the feature is at least partially made superfluous by the shared subscription feature which is part of the MQTT 5.0 spec and which we implemented for both MQTT 5.0 but also for MQTT 3.1.1 and 3.1. Shared subscriptions are a standard way of distributing messages and load balancing them between subscribers. Shared subscriptions are also more scalable than using the `allow_multiple_sessions` feature to distribute messages between clients as each shared subscriber has its own message queue while in the `allow_multiple_sessions` case a single message queue is shared.

The `allow_multiple_sessions` feature, however, is non-trivial, and there are many edge cases which can be hard to understand, especially when using it in a clustered setting. Indeed there a quite a few issues here on github concerning the feature. The feature also complicates the code base and tests quite a bit - and if we end up removing the code we would be able to simplify the code base significantly.

But before making any decisions we'd like some feedback from the community. Are you using the feature? If so, for what and could one perhaps conceive of other ways of solving the problem without using `allow_multiple_sessions`?

Thanks!

**Comment by cars10w** [comment]
Hi Lars,
we are not going to use that feature. In the case that we need something like that in the future, we would rather opt for the standardized thing (aka shared subs) and make use of that.

You already have outlined the drawbacks of `allow_multiple_sessions` as compared to the shared subscriptions, and therefore it sounds very reasonable to remove it.

**TL;DR** let's do it :-)

**Comment by codeadict** [comment]
We used it and moved to shared subscriptions. I'm always up for the standard way and i have experienced that setting being a bit problematic using a cluster.

**Comment by larshesel** [comment]
We've now marked `allow_multiple_sessions` as deprecated (as of 1.8.0) and added a new config option which is called `coordinate_registrations` which means clients connected with the same cilent-id will eventually be resolved and the oldest disconnected. This behaviour was a side-effect of `allow_multiple_sessions` and is desired as it is much faster than insisting on strict coordination all the time to strictly satisfy the MQTT spec.
